src/dsa/rpo_falcon512/signature.rs

```diff
@@ -42,245 +44,344 @@ use crate::utils::string::*;
 /// 3. 625 bytes encoding the `s2` polynomial above.
 ///
 /// The total size of the signature (including the extended public key) is 1563 bytes.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct Signature {
-    pub(super) pk: PublicKeyBytes,
-    pub(super) sig: SignatureBytes,
-
-    // Cached polynomial decoding for public key and signatures
-    pub(super) pk_polynomial: OnceCell<Polynomial>,
-    pub(super) sig_polynomial: OnceCell<Polynomial>,
+    header: SignatureHeader,
+    s2: SignaturePoly,
+    nonce: Nonce,
+    h: PubKeyPoly,
+    htp: HashToPoint,
```

Comment:
This should work. Should I make the required changes to implement it?

Reply:
I was thinking about this some more and now wondering if we are trying to do too much. Specifically, trying to support both signatures (canonical and our variant) publicly and with a well-structured API is going to be difficult (or more difficult than what we need). Even if we make signature encoding general as I suggested above, we'll still need to:

1. Figure out to to sometimes pass the message as `Word` and sometimes as array of bytes to the `sign()` method (as you've mentioned in one of your comments).
2. Figure out how to make s `verify()` function to work sometimes the a public key commitment and other times with the full public key.

Doing the above generically will be quite complicated (and may clutter the interfaces quite a bit). So, maybe we should go back to exporting only `RpoFalcon512` signature and keep the canonical variant internal and used for testing purposes only.

One way to do this is to create "helper" functions for signing and verification. For example, signing helper could look like this:
```Rust
pub fn sign_helper<R: Rng>(
    &SecretKey,
    c: Polynomial<FalconFelt>, // we could also created a dedicated type MessagePoly
    rng: &mut R,
) -> Result<(SignaturePoly, SignaturePoly), FalconError>
```

The caller of this function would be required to generate the nonce and compute hash-to-point before calling this function. The function would return the tuple of `(s1, s2)` polynomials.

Then, the `PrivateKey::sign()` method could look like this:
```Rust
pub fn sign<R: Rng>(
    &self,
    message: Word,
    rng: &mut R,
) -> Result<Signature, FalconError> {
    // generate nonce
    // call the RPO version of hash-to-point
    // call sign_helper
    // build the signature from (s1, s2)
}
```

For testing purposes, we could also create `sign_canonical()` function which could have the following signature:
```Rust
fn sign_canonical<R: Rng>(
    &self,
    message: &[u8],
    rng: &mut R,
) -> Result<CanonicalSignature, FalconError>
```

For verification, the helper could look as follows:
```Rust
pub fn verify_helper(c: MessagePoly, s2: SignaturePoly, h: PubKeyPoly) -> bool
```
And then this method would be used similarly to the signature helper by the Rpo Falcon as well as canonical Falcon.

Btw, one reason why signature helper returns `(s1, s2)` is that I think it might be good to implement RPO signature in "key recovery" mode. This will allow us to use just the secret key to generate signatures and then we can get rid of the `KeyPair` struct (also, as a nice bonus, signatures should be 15% smaller).

We can still keep our current MASM implementation of signature verification - we just need to recover the key from the signature first. My understating is that the main challnage with this approach is to ensure that `s2` is invertible - and I'm not sure how complicated this is.

Reply:
Makes sense. I have now reverted to the previous way of implementing the hash-to-point logic. 
I have also implemented the key recovery mode.